File: java/semantickernel-core-skills/src/main/java/com/microsoft/semantickernel/coreskills/WaitSkill.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.coreskills;

import com.microsoft.semantickernel.skilldefinition.annotations.DefineSKFunction;

import reactor.core.publisher.Mono;

import java.time.Duration;

/** WaitSkill provides a set of functions to wait before making the rest of operations. */
public class WaitSkill {

    /**
     * Wait for a certain number of seconds.
     *
     * <p>Examples:
     *
     * <p>SKContext context = SKBuilders.context().build();
     *
     * <p>context.setVariable("input","10");
     *
     * <p>{{wait.seconds $input}}
     *
     * @param seconds The number of seconds to wait as a string.
     * @return A Mono that completes after the specified delay.
     * @throws IllegalArgumentException If the input is not a valid number.
     */
    @DefineSKFunction(name = "seconds", description = "Wait a given amount of seconds")
    public Mono<Void> wait(String seconds) {
        try {
            double sec = Math.max(Double.parseDouble(seconds), 0);
            long milliseconds = (long) (sec * 1000);
            return Mono.delay(Duration.ofMillis(milliseconds)).then();
        } catch (NumberFormatException e) {
            return Mono.error(new IllegalArgumentException("seconds text must be a number", e));
        }
    }
}

File: python/semantic_kernel/core_skills/wait_skill.py
# Copyright (c) Microsoft. All rights reserved.

import asyncio

from semantic_kernel.sk_pydantic import PydanticField
from semantic_kernel.skill_definition import sk_function


class WaitSkill(PydanticField):
    """
    WaitSkill provides a set of functions to wait for a certain amount of time.

    Usage:
        kernel.import_skill(WaitSkill(), skill_name="wait")

    Examples:
        {{wait.seconds 5}} => Wait for 5 seconds
    """

    @sk_function(description="Wait for a certain number of seconds.")
    async def wait(self, seconds_text: str):
        try:
            seconds = max(float(seconds_text), 0)
        except ValueError:
            raise ValueError("seconds text must be a number")
        await asyncio.sleep(seconds)
